ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,412 @@
+  class PaymentNetworkError < Temporal::RetryableError
+    attr_reader :status
+
+    def initialize(message, status)
+      super(message)
+      @status = status
+    end
+  end
+
+  def execute(account, amount)
+    if amount < balance_of(account.number)
+      raise InsufficientFunds, "Account #{account.number} does not have enough funds"
+    end
+
+    response = process_payment(account.number, amount)
+    if response.failure?
+      raise PaymentNetworkError, response.message, response.status
+    end
+
+    ...
+  end
+end
+
+class RenewSubscriptionWorkflow < Temporal::Workflow
+  def execute
+    ...
+
+    begin
+      workflow.execute_activity(ProcessPaymentActivity, account, amount)
+    rescue Temporal::ActivityError => error
+      if error.cause.is_a?(Temporal::ApplicationError)
+        case error.cause.original_class
+        when 'ProcessPaymentActivity::InsufficientFunds'
+          # handle insufficient funds error here
+        when 'ProcessPaymentActivity::PaymentNetworkError'
+          # handle payment network error
+          # notice that status is not available
+        end
+      end
+
+      # You'll need to re-raise the error in case it wasn't handled
+      raise
+    end
+
+    ...
+  end
+end
+```
+
+Application errors will be encoded/decoded using the same Payload converters as the other payloads.
+This also means that if an error class implements `#to_json` / `.json_create` (as suggested by the
+[Ruby's JSON module](https://ruby-doc.org/stdlib-3.1.2/libdoc/json/rdoc/JSON.html#module-JSON-label-Custom+JSON+Additions))
+then we will be able to recreate the original error class:
+
+```ruby
+  class ProcessPaymentActivity < Temporal::Activity
+    class InsufficientFunds < Temporal::NonRetryableError
+      # Add default serialization to an standard looking error
+      include Temporal::JSON::SerializableError
+    end
+    class PaymentNetworkError < Temporal::RetryableError
+      attr_reader :status
+
+      def self.json_create(object)
+        new(object['message'], object['status'])
+      end
+
+      def initialize(message, status)
+        super(message)
+        @status = status
+      end
+
+      def to_json(*args)
+        {
+          JSON.create_id  => self.class.name,
+          'message' => message,
+          'status' => 'status',
+        }.to_json(*args)
+      end
+    end
+
+    # same #execute as the previous example
+  end
+
+  class RenewSubscriptionWorkflow < Temporal::Workflow
+    def execute
+      ...
+
+      begin
+        workflow.execute_activity(ProcessPaymentActivity, account, amount)
+      rescue ProcessPaymentActivity::InsufficientFunds
+        # handle insufficient funds error here
+      rescue ProcessPaymentActivity::PaymentNetworkError => e
+        # handle payment network error
+        # e.status is available as expected
+      end
+
+      ...
+    end
+  end
+```
+
+To allow customisation of this behaviour we will expose an interface for failure conversion:
```

Comment:
I think this should be the only place we allow customization

Reply:
It is the only place. Unless you're referring to JSON additions, but you can't possibly expect users to manually convert every possible ApplicationFailure `type` in their custom failure converter

Reply:
> Unless you're referring to JSON additions, but you can't possibly expect users to manually convert every possible ApplicationFailure type in their custom failure converter

I am referring to that. We should not auto-deserialize to user-defined errors in Ruby until we have a general discussion about how that works in SDKs. Reusing the "type" to mean "class" and "error details" to mean "class info" would be a new thing to Temporal.

Reply:
Agreed to remove this from this proposal until there's a cross-SDK solution